app/api/db/__init__.py

```diff
@@ -1,60 +1,159 @@
+
+    This class is used to manage database connections for the Flask app.
+    It has methods for getting a new connection or session object.
+    """
+
+    _engine: sqlalchemy.engine.Engine
+
+    def __init__(self) -> None:
+        self._engine = _create_db_engine()
+
+    def init_app(self, app: Flask) -> None:
+        """Initialize the Flask app.
+
+        Add the database to the Flask app's extensions so that it can be
+        accessed by request handlers using the current app context.
+
+        see get_db
+        """
+        app.extensions[FLASK_EXTENSION_KEY] = self
 
-    if check_migrations_current:
-        have_all_migrations_run(engine)
+    def get_connection(self) -> Connection:
+        """Return a new database connection object.
 
-    engine.dispose()
+        Use the connection to execute SQL queries without using the ORM.
 
-    return session_factory
+        Usage:
+            with db.get_connection() as conn:
+                conn.execute(...)
+        """
+        return self._engine.connect()
+
+    def get_session(self) -> Session:
+        """Return a new session object.
+
+        In general, only one session object should be created per request.
+
+        If you want to automatically commit or rollback the session, use
+        the session.begin() context manager.
+        See https://docs.sqlalchemy.org/en/13/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it
+
+        Example:
+            with db.get_session() as session:
+                with session.begin():
+                    session.add(...)
+                # session is automatically committed here
+                # or rolled back if an exception is raised
+        """
+        return Session(bind=self._engine, expire_on_commit=False, autocommit=False)
+
+    def check_db_connection(self) -> None:
+        """Check that we can connect to the database and log some info about the connection."""
+        logger.info("connecting to postgres db")
+        with self.get_connection() as conn:
+            conn_info = conn.connection.dbapi_connection.info  # type: ignore
+
+            logger.info(
+                "connected to postgres db",
+                extra={
+                    "dbname": conn_info.dbname,
+                    "user": conn_info.user,
+                    "host": conn_info.host,
+                    "port": conn_info.port,
+                    "options": conn_info.options,
+                    "dsn_parameters": conn_info.dsn_parameters,
+                    "protocol_version": conn_info.protocol_version,
+                    "server_version": conn_info.server_version,
+                },
+            )
+            verify_ssl(conn_info)
+
+            # TODO add check_migrations_current to config
+            # if check_migrations_current:
+            #     have_all_migrations_run(engine)
+
+
+def init() -> DB:
+    return DB()
+
+
+def get_db(app: Flask) -> DB:
```

Comment:
my understanding is that while flask is mostly for the web use case, that it also supports custom CLI commands (https://flask.palletsprojects.com/en/2.2.x/cli/#custom-commands) that appears to be the same as what we're doing manually with the individual script files in the scripts directory, so it'd be nice to consolidate and use the flask app as the central "app" that we run in both the "web app" and "batch job" contexts. thoughts?

Reply:
Another advantage with using the Flask custom CLI command framework rather than making our homegrown entry point is that we'd be able to leverage blueprints. In the future I'd like us to build reusable blueprints for things like "doc upload endpoints" or "user account management" (change address, change name, etc), and those blueprints could come with their own CLI commands like "disable user account".

Reply:
Wasn't aware that Flask supported CLI commands like that, seems a little odd, but I do like click over something like argparse.

Just brainstorming, what flask features would we be able to leverage with running via CLI? I know flask manages several contexts like a request context.

I'd be curious to see some scripts built out using it as those examples are very basic. Something similar to the ECS tasks on Mass (at least the entry-point). I do think having some utilities/patterns for how to build a task, and instrument them would be useful. Porting over something like the `background_task` from the Mass project which sets up logging, metrics, task name, and so on.

Reply:
brainstorming some features that we can use:

* application_context, like you suggested, which can be used to manage extensions that we install in the future, and custom extension like the DB client we added in this PR. so CLI scripts can do `db = api.db.get_db(current_app)`
* [debug mode](https://flask.palletsprojects.com/en/2.2.x/cli/#debug-mode) interactive debugger and reloader
* [registering CLI commands with blueprints](https://flask.palletsprojects.com/en/2.2.x/cli/#registering-commands-with-blueprints), like i described in the previous comment
* looks like some [flask extensions/plugins could add their own commands](https://flask.palletsprojects.com/en/2.2.x/cli/#plugins)

it also looks like we could still [build custom scripts using click](https://flask.palletsprojects.com/en/2.2.x/cli/#custom-scripts) and pass it the create_app factory if we wanted to go that route

> I'd be curious to see some scripts built out using it as those examples are very basic. Something similar to the ECS tasks on Mass (at least the entry-point). I do think having some utilities/patterns for how to build a task, and instrument them would be useful. Porting over something like the background_task from the Mass project which sets up logging, metrics, task name, and so on.

are you thinking we should do that in this PR (knowing that i might have broken the create_user_csv script) or would you be open to doing that in a future PR since this PR is not small